ClusterStorageContainer with `kit://` prefix and Kserve not working

**Use case:**
Our idea is to use KitOps to push model files into OCI compatible repository in AWS ECR and use that model for serving in Kserve EKS. Kserve itself does support ECR (as `oci://` prefix in combination with supported model serving runtimes) but only when used as container repo.

**What I have checked**
I found that there is part of KitOps repo when something similar is already covered here: https://github.com/kitops-ml/kitops/tree/main/build/dockerfiles/KServe. 

First of I think that there is a typo in image path: https://github.com/kitops-ml/kitops/blob/98d3e1e1c29ae86a14446f89eec77d27d0fc5c22/build/dockerfiles/KServe/Dockerfile#L1 because it is pointing to image that it is not reachable:

```
ERROR: failed to solve: ghcr.io/kitops-ml/kit:next: failed to resolve source metadata for 
ghcr.io/kitops-ml/kit:next: failed to authorize: failed to fetch anonymous token: 
unexpected status from GET request to https://ghcr.io/token?scope=repository%3Akitops-ml%2Fkit%3Apull&service=ghcr.io: 
403 Forbidden

docker pull ghcr.io/kitops-ml/kit:next
Error response from daemon: Head "https://ghcr.io/v2/kitops-ml/kit/manifests/next": denied
```

Should that point to `ghcr.io/kitops-ml/kitops:next` like your init container?
https://github.com/kitops-ml/kitops/blob/98d3e1e1c29ae86a14446f89eec77d27d0fc5c22/build/dockerfiles/init/Dockerfile#L3

Assuming the answer is yes, I've tried following further your guide in readme by creating `ClusterStorageContainer` and `InferenceService` that uses `kit://` prefix:

```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: sklearn-movies-kitops-custom-storage
  namespace: kserve
spec:
  predictor:
    model:
      modelFormat:
        name: sklearn
      storageUri: "kit://<ACCOUNT_ID>.dkr.ecr.eu-central-1.amazonaws.com/kserve/models:kitpacked-model-1"
```

**Issue**
Upon deploying inference manifest, there are few resources created, one of which is ReplicaSet that below log is from:
```
'Pod "sklearn-movies-kitops-custom-storage-predictor-856bfd8966-hklgn"
      is invalid: [spec.volumes[3].name: Duplicate value: "kserve-provision-location",
      spec.containers[0].volumeMounts[3].mountPath: Invalid value: "/mnt/models":
      must be unique, spec.initContainers[1].name: Duplicate value: "kitops-storage-init"]'
    reason: FailedCreate
```
There is no container spawned, thus there is no log from it nor manifest used to create pod.
I'm guessing that I miss something out here or there is some underlying issue with example that you've provided. When did you test this integration? What was the version of KitOps and Kserve? Could you please take a look from your end?

**To Reproduce**
Steps to reproduce the behavior:
1. Follow KitOps Kserve example
2. Push model to ECR using `kit pack` and `kit push` commands
3. Deploy example inference manifest
4. Check `ReplicaSet` for error


**Version**
- KitOps: `next`, `v1.3.0`
- Kserve: `v0.14.1`
- K8s: `v1.31.3`

I'm guessing that you mean `spec.container.name` and not `metadata.name` - I can confirm that changing `spec.container.name` does make a difference! Init container is deployed and logs can be read.

So now my follow up question: how do you handle auth while using `kit://`? 
Using Kserve `enableModelcar` flag to enable OCI repo support (container based) `oci://` prefix is properly working with ECR, while what I see when using `kit://` prefix is:

- unpack flags: "-vv"
```
[INFO ] Unpacking to /mnt/models
[TRACE] HEAD https://<ACCOUNT_ID>.dkr.ecr.eu-central-1.amazonaws.com/v2/kserve/models/manifests/kitpacked-model-1 -> ERROR -- duration 14.01 ms
[ERROR] Failed to find reference <ACCOUNT_ID>.dkr.ecr.eu-central-1.amazonaws.com/kserve/models:kitpacked-model-1: unexpected error retrieving reference from remote: HEAD "https://<ACCOUNT_ID>.dkr.ecr.eu-central-1.amazonaws.com/v2/kserve/models/manifests/kitpacked-model-1": basic credential not found
```

- unpack flags: "-vv --plain-http"
```
[INFO ] Unpacking to /mnt/models
[TRACE] HEAD http://<ACCOUNT_ID>.dkr.ecr.eu-central-1.amazonaws.com/v2/kserve/models/manifests/kitpacked-model-1 -> ERROR -- duration 186793.19 ms
[ERROR] Failed to find reference <ACCOUNT_ID>.dkr.ecr.eu-central-1.amazonaws.com/kserve/models:kitpacked-model-1: unexpected error retrieving reference from remote: Head "http://<ACCOUNT_ID>.dkr.ecr.eu-central-1.amazonaws.com/v2/kserve/models/manifests/kitpacked-model-1": dial tcp 172.31.177.32:80: i/o timeout
```

is 
```
--cert string       Path to client certificate used for authentication (can also be set via environment variable KITOPS_CLIENT_CERT)
--key string        Path to client certificate key used for authentication (can also be set via environment variable KITOPS_CLIENT_KEY)
```
only option here? 
What I'm basically searching for is token based auth, which would enable possibility to leverage AWS IRSA.

--- a/.github/workflows/next-container-build.yaml
+++ b/.github/workflows/next-container-build.yaml
@@ -11,6 +11,7 @@ env:
   REGISTRY: ghcr.io
   IMAGE_NAME: ${{ github.repository }}
   INIT_IMAGE_NAME: ${{ github.repository }}-init
+  KIT_SERVE_IMAGE: ${{ github.repository }}-kserve
   NEXT_TAG: next
 
 permissions:
@@ -72,17 +73,20 @@ jobs:
             index:org.opencontainers.image.description=Kit CLI init container
             index:org.opencontainers.image.source=https://github.com/kitops-ml/kitops
             index:org.opencontainers.image.licenses=Apache-2.0
-
-      # - name: Generate artifact attestation for base container
-      #   uses: actions/attest-build-provenance@6149ea5740be74af77f260b9db67e633f6b0a9a1   # v1.4.2
-      #   with:
-      #     subject-name: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
-      #     subject-digest: ${{ steps.build-kit-container.outputs.digest }}
-      #     push-to-registry: true
-
-      # - name: Generate artifact attestation for base container
-      #   uses: actions/attest-build-provenance@6149ea5740be74af77f260b9db67e633f6b0a9a1   # v1.4.2
-      #   with:
-      #     subject-name: ${{ env.REGISTRY }}/${{ env.INIT_IMAGE_NAME }}
-      #     subject-digest: ${{ steps.build-kit-init-container.outputs.digest }}
-      #     push-to-registry: true
+     
+      - name: Build and push Kit KServe container
+        id: build-kit-serve-container
+        uses: docker/build-push-action@471d1dc4e07e5cdedd4c2171150001c434f0b7a4    # v6.15.0
+        with:
+          platforms: linux/amd64,linux/arm64
+          push: true
+          context: build/dockerfiles/KServe
+          file: build/dockerfiles/KServe/Dockerfile
+          build-args: |
+            KIT_BASE_IMAGE=${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}@${{ steps.build-kit-container.outputs.digest }}
+          tags: |
+            ${{ env.REGISTRY }}/${{ env.KIT_SERVE_IMAGE }}:${{ env.NEXT_TAG }}
+          annotations: |
+            index:org.opencontainers.image.description=KitOps KServe container
+            index:org.opencontainers.image.source=https://github.com/kitops-ml/kitops
+            index:org.opencontainers.image.licenses=Apache-2.0

--- a/.github/workflows/platform-release.yaml
+++ b/.github/workflows/platform-release.yaml
@@ -20,6 +20,7 @@ env:
   REGISTRY: ghcr.io
   IMAGE_NAME: ${{ github.repository }}
   INIT_IMAGE_NAME: ${{ github.repository }}-init
+  KIT_SERVE_IMAGE: ${{ github.repository }}-kserve
 
 permissions:
   contents: write
@@ -394,6 +395,24 @@ jobs:
             index:org.opencontainers.image.source=https://github.com/kitops-ml/kitops
             index:org.opencontainers.image.licenses=Apache-2.0
 
+      - name: Build and push Kit KServe container
+        id: build-kit-serve-container
+        uses: docker/build-push-action@471d1dc4e07e5cdedd4c2171150001c434f0b7a4    # v6.15.0
+        with:
+          platforms: linux/amd64,linux/arm64
+          push: true
+          context: build/dockerfiles/KServe
+          file: build/dockerfiles/KServe/Dockerfile
+          build-args: |
+            KIT_BASE_IMAGE=${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}@${{ steps.build-kit-container.outputs.digest }}
+          tags: |
+            ${{ env.REGISTRY }}/${{ env.KIT_SERVE_IMAGE }}:latest
+            ${{ env.REGISTRY }}/${{ env.KIT_SERVE_IMAGE }}:${{ github.ref_name }}
+          annotations: |
+            index:org.opencontainers.image.description=Kit KServe container
+            index:org.opencontainers.image.source=https://github.com/kitops-ml/kitops
+            index:org.opencontainers.image.licenses=Apache-2.0
+
       - name: Generate artifact attestation for base container
         uses: actions/attest-build-provenance@c074443f1aee8d4aeeae555aebba3282517141b2   # v2.2.3
         with:
@@ -401,9 +420,16 @@ jobs:
           subject-digest: ${{ steps.build-kit-container.outputs.digest }}
           push-to-registry: true
 
-      - name: Generate artifact attestation for base container
+      - name: Generate artifact attestation for kit init container
         uses: actions/attest-build-provenance@c074443f1aee8d4aeeae555aebba3282517141b2   # v2.2.3
         with:
           subject-name: ${{ env.REGISTRY }}/${{ env.INIT_IMAGE_NAME }}
           subject-digest: ${{ steps.build-kit-init-container.outputs.digest }}
+          push-to-registry: true
+
+      - name: Generate artifact attestation for kit kserve container
+        uses: actions/attest-build-provenance@c074443f1aee8d4aeeae555aebba3282517141b2   # v2.2.3
+        with:
+          subject-name: ${{ env.REGISTRY }}/${{ env.KIT_SERVE_IMAGE }}
+          subject-digest: ${{ steps.build-kit-serve-container.outputs.digest }}
           push-to-registry: true

--- a/build/dockerfiles/KServe/README.md
+++ b/build/dockerfiles/KServe/README.md
@@ -1,23 +1,13 @@
-# Kitops ClusterStorageContainer image for KServe
+# KitOps ClusterStorageContainer image for KServe
 
 The Dockerfile in this directory is used to build an image that can run as a ClusterStorageContainer for [KServe](https://kserve.github.io/website/master/).
 
-## Building
-To build the image, `docker` or `podman` is required. From the root of this repository, set the `$KIT_KSERVE_IMAGE` to the image tag you want to build and run
-```bash
-docker build -t $KIT_KSERVE_IMAGE .
-```
-
-By default, the image will be built using `ghcr.io/kitops-ml/kitops:next` as a base. This can be overridden (to build using a specific version of Kit, for example) by using the build arg `KIT_BASE_IMAGE`:
-```shell
-# Build the image based on Kit v1.3.0 instead of 'next'
-docker build -t kit-init-container:latest --build-arg KIT_BASE_IMAGE=ghcr.io/kitops-ml/kitops:v1.3.0 .
-```
-
 ## Installing
-The following process will create a new ClusterStorageContainer that uses `kit` to support KServe InferenceServices with storage URIs that have the `kit://` prefix.
 
-Create the following ClusterStorageContainer custom resource in a Kubernetes cluster with KServe installed
+The following process creates a new ClusterStorageContainer that uses `kit` to support KServe InferenceServices with storage URIs that have the `kit://` prefix.
+
+Create the following `ClusterStorageContainer` custom resource in a Kubernetes cluster with KServe installed. Note that the sample below uses the `ghcr.io/kitops-ml/kitops-kserve:next` image, although the repository includes other tags you can use.
+
 ```yaml
 apiVersion: "serving.kserve.io/v1alpha1"
 kind: ClusterStorageContainer
@@ -25,8 +15,8 @@ metadata:
   name: kitops
 spec:
   container:
-    name: kit-storage-initializer
-    image: $KIT_KSERVE_IMAGE
+    name: storage-initializer
+    image: ghcr.io/kitops-ml/kitops-kserve:next
     imagePullPolicy: Always
     env:
       - name: KIT_UNPACK_FLAGS
@@ -41,13 +31,40 @@ spec:
     - prefix: kit://
 ```
 
-Once this CR is installed, modelkits can be used in InferenceServices by specifying with the `kit://` URI:
+Once this custom resource is installed, ModelKits can be used in InferenceServices by specifying the ModelKit URI with the `kit://` prefix in the `storageUri` field:
+
 ```yaml
-storageUri: kit://<modelkit-reference>
+apiVersion: "serving.kserve.io/v1beta1"
+kind: "InferenceService"
+metadata:
+  name: "iris-model"
+spec:
+  predictor:
+    model:
+      modelFormat:
+        name: sklearn
+      storageUri: kit://<modelkit-reference>
+```
+
+## Building
+
+To build the image, `docker` or `podman` is required. From the root of this repository, set the `$KIT_KSERVE_IMAGE`  environment variable to the image tag you want to build and run
+
+```bash
+docker build -t $KIT_KSERVE_IMAGE .
+```
+
+By default, the image will be built using `ghcr.io/kitops-ml/kitops:next` as a base. This can be overridden by specifying the build argument `KIT_BASE_IMAGE` to use a specific version of Kit. For example:
+
+```shell
+# Build the image based on Kit v1.3.0 instead of 'next'
+docker build -t kitops-kserve:latest --build-arg KIT_BASE_IMAGE=ghcr.io/kitops-ml/kitops:v1.3.0 .
 ```
 
 ## Configuration
-The Kit KServe container supports specifying additional flags as supported by the `kit unpack` command. Additional flags are read from the `KIT_UNPACK_FLAGS` environment variable in the ClusterStorageContainer. For example, the following adds `-v` and `--plain-http` for all unpack commands:
+
+The Kit KServe container supports specifying additional flags for the `kit unpack` command. These flags are read from the KIT_UNPACK_FLAGS environment variable in the ClusterStorageContainer. For example, the following configuration adds `-v` and `--plain-http` for all unpack commands:
+
 ```yaml
     env:
       - name: KIT_UNPACK_FLAGS
@@ -55,4 +72,5 @@ The Kit KServe container supports specifying additional flags as supported by th
 ```
 
 ## Additional links
+
 * [KServe ClusterStorageContainer documentation](https://kserve.github.io/website/master/modelserving/storage/storagecontainers/)

